controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -99,11 +137,481 @@ func (r *StorageClassClaimReconciler) reconcileConsumerPhases(ctx context.Contex
+				"osd": "allow rw tag cephfs *=*",
+				"mds": fmt.Sprintf("allow rw path=/volumes/%s", r.cephFilesystemSubVolumeGroup.Name),
+			},
+		}
+		return nil
+	})
+
+	if err != nil {
+		r.log.Error(
+			err,
+			"Failed to update CephClient.",
+			"CephClient",
+			klog.KRef(r.cephClientCephFSNode.Namespace, r.cephClientCephFSNode.Name),
+		)
+		return err
+	}
+
+	phase := ""
+	if r.cephClientCephFSNode.Status != nil {
+		phase = string(r.cephClientCephFSNode.Status.Phase)
+	}
+
+	r.setCephResourceStatus(r.cephClientCephFSNode.Name, "CephClient", phase, nil)
+
+	return nil
+}
+
+func (r *StorageClassClaimReconciler) verifyCephResourcesDoNotExist() bool {
+	for _, cephResource := range r.storageClassClaim.Status.CephResources {
+		switch cephResource.Kind {
+		case "CephClient":
+			cephClient := &rookCephv1.CephClient{}
+			cephClient.Name = cephResource.Name
+			cephClient.Namespace = r.namespace
+			err := r.get(cephClient)
+			cephUserType := cephClient.Annotations[controllers.StorageCephUserTypeAnnotation]
+			if cephUserType != "healthchecker" && (err == nil || !errors.IsNotFound(err)) {
+				return false
+			}
+		case "CephBlockPool":
+			cephBlockPool := &rookCephv1.CephBlockPool{}
+			cephBlockPool.Name = cephResource.Name
+			cephBlockPool.Namespace = r.namespace
+			if err := r.get(cephBlockPool); err == nil || !errors.IsNotFound(err) {
+				return false
+			}
+		case "CephFilesystemSubVolumeGroup":
+			cephFilesystemSubVolumeGroup := &rookCephv1.CephFilesystemSubVolumeGroup{}
+			cephFilesystemSubVolumeGroup.Name = cephResource.Name
+			cephFilesystemSubVolumeGroup.Namespace = r.namespace
+			if err := r.get(cephFilesystemSubVolumeGroup); err == nil || !errors.IsNotFound(err) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (r *StorageClassClaimReconciler) setCephResourceStatus(name string, kind string, phase string, cephClients map[string]string) {
+
+	cephResourceSpec := r.cephResourcesByName[name]
+
+	if cephResourceSpec == nil {
+		cephResourceSpec = &ocsv1alpha1.CephResourcesSpec{
+			Name:        name,
+			Kind:        kind,
+			CephClients: cephClients,
+		}
+		r.storageClassClaim.Status.CephResources = append(r.storageClassClaim.Status.CephResources, cephResourceSpec)
+		r.cephResourcesByName[name] = cephResourceSpec
+	}
+
+	cephResourceSpec.Phase = phase
+}
+
+func addStorageRelatedAnnotations(obj client.Object, storageClassClaimName, storageClaim, cephUserType string) {
+	annotations := obj.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+		obj.SetAnnotations(annotations)
+	}
+
+	annotations[StorageClassClaimAnnotation] = storageClassClaimName
+	annotations[controllers.StorageClaimAnnotation] = storageClaim
+	annotations[controllers.StorageCephUserTypeAnnotation] = cephUserType
+}
+
+// Checks whether a string is contained within a slice
+func contains(slice []string, s string) bool {
+	for _, item := range slice {
+		if item == s {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
Please reuse this function from other places. Do not duplicate it.

Reply:
done